### PR TITLE
Adopt linrtunner as the linting tool

### PR DESCRIPTION
**Pull request by justinchuby**
### Description

`lintrunner` is a linter runner successfully used by pytorch, onnx and onnx-script. This PR adopts `lintrunner` to onnxruntime and hand fixed ~2000 flake8 errors in Python code.  `lintrunner` now runs all required python lints including `flake8`, `black` and `isort`. Future lints like `clang-format` can be added.

In particular, this PR **removed some very suboptimal patterns**:

- `import *`
- mutable values as default in function definitions (`def func(a=[])`)
- `not xxx in` -> `xxx not in` membership checks
- bare excepts (`except:` -> `except Exception`)
- unused imports
- unused local variables

The added workflow supports SARIF code scanning reports on github, example snapshot:

image

### Motivation and Context

Unified linting experience in CI and local.

### CODE REVIEW

**Review by mszhanyi — APPROVED** [review]
thanks.

**Comment by justinchuby** [comment]
Todo: the rest of the lints; ort training import order

File: orttraining/orttraining/eager/opgen/opgen/parser.py [hunk 1]
```diff
@@ -1,9 +1,10 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
-from opgen.lexer import *
+from typing import List, Optional, Tuple, Union
+
 from opgen.ast import *
-from typing import List, Tuple, Union, Optional
+from opgen.lexer import *
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply orttraining/orttraining/eager/opgen/opgen/parser.py:7]: Existing error I didn't have the bandwidth fixing

**Comment by justinchuby** [comment]
test_ort_init: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=864045&view=logs&jobId=420dfd90-b28e-514e-c751-f2d154829d2b&j=420dfd90-b28e-514e-c751-f2d154829d2b&t=4d5dca6a-78a0-5ee0-89a5-1142a9a98332

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 2]
```diff
@@ -1028,7 +1027,7 @@ def _infer_GatherND(self, node):
         data_shape = self._get_shape(node, 0)
         data_rank = len(data_shape)
         indices_shape = self._get_shape(node, 1)
-        indices_rank = len(indices_shape)
+        len(indices_shape)
```
Comment (edgchen1) [inline onnxruntime/python/tools/symbolic_shape_infer.py:1030]: do we need this line?

File: .lintrunner.toml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    # ignore generated flatbuffers code
+    'onnxruntime/core/flatbuffers/ort_flatbuffers_py/**',
+    # TODO enable
+    './onnxruntime/python/tools/**',
+    # FIXME(#7032): ignore test code for now
+    'onnxruntime/test/**',
+    # TODO enable
+    'orttraining/**',
+    # FIXME(#7032): ignore server code for now
+    'server/**',
```
Comment (edgchen1) [inline .lintrunner.toml:55]: server/ has been removed

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 2 above)
Comment (justinchuby) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1030]: Good catch thanks

**Comment by justinchuby** [comment]
I'll be back at this after the release to cause less disruption

**Comment by justinchuby** [comment]
Replaced by #15085

**Pull request closed without merging**